Make team kills condensed in killfeed

VoxelDoesCode opened the pull request:
For those who still have the killfeed on, I wanted to modernize it, by having team kills become one line instead of a bunch!

![image](https://user-images.githubusercontent.com/95713843/226060109-ed74c2f6-1289-4247-81dc-1725428ddf59.png)
## Checklist

- [x] Tested the change ingame
- [x] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

VoxelDoesCode committed:
Make team kills condensed in killfeed

Clang + Formatting test before merge

def- commented:
Looks cool!

fokkonaut commented:
Shouldnt strong be left? iirc the guy who killed first gets strong

fokkonaut commented:
I would probably do this a little different on the networking side tho. Maybe a SvKillMsgTeam which simply gets an integer for the Team. The client knows about teams states and could render a kill msg for the whole team accordingly, with all skins. No need to send multiple messages then, and also no need to send the original + new messages.

VoxelDoesCode commented:
But when teams die, all members currently send a message. Are you implying I do all of the logic inside killmessage.cpp, while still keeping the `Single` bool inside the `Die()` function?

VoxelDoesCode committed:
Simplify` KillMsgTeams` NetMsg, fixing bugs

Clang format before squash

Split apart OnMessage() before squash

forgot something

Robyt3 commented on src/game/client/components/killmessages.cpp:
```diff
@@ -103,6 +105,80 @@ void CKillMessages::OnMessage(int MsgType, void *pRawMsg)
+				}
+			}
+
+			char aBuf[8];
+			str_format(aBuf, sizeof(aBuf), "Team %d", Kill.m_VictimDDTeam);
+			str_copy(Kill.m_aVictimName, aBuf);
```
`aBuf` is not necessary, you can do `str_format(Kill.m_aVictimName, ...)` directly.

Chairn commented on src/game/client/components/killmessages.cpp:
```diff
@@ -112,13 +127,28 @@ void CKillMessages::OnMessage(int MsgType, void *pRawMsg)
+			if(Kill.m_TeamSize != 1)
+			{
+				char aBuf[8];
+				str_format(aBuf, sizeof(aBuf), "Team %d", Kill.m_VictimDDTeam);
```
You can directly format m_aVictimName

VoxelDoesCode committed:
Bring Team logic into teams.cpp

MsgPlus -> Msg

fokkonaut commented:
> 

Either that, yea, or a separated CNetMsg_KillMsgTeam.
That could simply hold a int Team so the client knows what to do.
I showed a little algorithm to Voxel which could sort the players of a specific team by strong weak id for the correct order.
I think having this separated message is worth it, so that it's still possible to send a normal kill msg even if someone is in team. This would also allow my server for example to send kill messages with a victim and a killer, even when they are in team. That happens when people are in 1vs1 for example.

VoxelDoesCode committed:
Simplify NetMsg even more, with help from Fokko

Quick changes

Add m_First

one more thing

fokko pointed this out at 3 am

Final push

Robyt3 commented on src/game/server/teams.cpp:
```diff
@@ -1137,4 +1142,4 @@ int CGameTeams::GetFirstEmptyTeam() const
 			return i;
 	return -1;
-}
+}
```
Leave newline at the end of the file.

Robyt3 commented on src/game/client/components/killmessages.h:
```diff
@@ -6,6 +6,11 @@
 
 #include <game/client/render.h>
 
+enum
+{
+	MAX_KILLMSGS = 5,
+	MAX_KILLMSGTEAM = 4,
+};
```
Can this `enum` be inside `CKillMessages`?

VoxelDoesCode replied:
I left it out, so that it could be accessed by `struct CKillMsg`. I wanted it to be easy to change the number of tees needed for the buffer without having to manually type it in.

Robyt3 replied:
It should still work the same if you move the enum to the top of the `CKillMessages` class, as `CKillMsg` is nested in `CKillMessages`.

Robyt3 commented on src/game/client/components/killmessages.cpp:
```diff
@@ -103,6 +108,86 @@ void CKillMessages::OnMessage(int MsgType, void *pRawMsg)
+			}
+		}
+
+		std::sort(vStrongWeakSorted.begin(), vStrongWeakSorted.end(), [](auto &left, auto &right) { return left.second > right.second; });
```
`Left` and `Right`

Robyt3 commented on src/game/client/components/killmessages.cpp:
```diff
@@ -103,6 +108,86 @@ void CKillMessages::OnMessage(int MsgType, void *pRawMsg)
+				Pair.first = i;
+				CCharacter *pChr = m_pClient->m_GameWorld.GetCharacterByID(i);
+				Pair.second = pChr ? pChr->GetStrongWeakID() : 0;
+				vStrongWeakSorted.push_back(Pair);
```
You could use `emplace_back(i, pChr ? pChr->GetStrongWeakID() : 0)` instead of manually creating the `std::pair`.

fokkonaut commented:
Could you rename the netmsg to KillMsgTeam? Seems more fitting imo

heinrich5991 commented on src/game/client/components/killmessages.cpp:
```diff
@@ -103,6 +107,83 @@ void CKillMessages::OnMessage(int MsgType, void *pRawMsg)
+			}
+		}
+
+		std::sort(vStrongWeakSorted.begin(), vStrongWeakSorted.end(), [](auto &Left, auto &Right) { return Left.second > Right.second; });
```
I don't see a reason for unstable sort, so I guess `std::stable_sort` would be better.

heinrich5991 commented on src/game/client/components/killmessages.h:
```diff
@@ -5,10 +5,14 @@
+	enum
+	{
+		MAX_KILLMSGS = 5,
+		MAX_KILLMSGTEAM = 4,
```
`MAX_KILLMSG_TEES`? `MAX_KILLMSG_TEAM_MEMBERS`? The current name didn't tell me it's about the number of tees displayed per team.

heinrich5991 commented on datasrc/network.py:
```diff
@@ -513,4 +513,9 @@
 	]),
+    
+	NetMessageEx("Sv_KillMsgTeam", "killmsgteam@netmsg.ddnet.tw", [
+		NetIntRange("m_Team", 0, 'MAX_CLIENTS'),
```
Should that be `MAX_CLIENTS-1`? Or is `MAX_CLIENTS` the "super" team? I forgot.

VoxelDoesCode replied:
Ah, forgot that.

fokkonaut replied:
MAX_CLIENTS is correct, it is TEAM_SUPER.

VoxelDoesCode replied:
I can revert this for future-proofing, if that's alright.

heinrich5991 reviewed (commented):
Just three small comments, looks good to me otherwise.

VoxelDoesCode commented:
I'm thinking about potentially adding in a `m_Killer` skin buffer alongside the `m_Victim` buffer, for vanilla related games. But I'll save that for another PR.

heinrich5991 commented:
bors r+

Thanks for working on these graphical improvements. :)
